### ISSUE TITLE
GD-487: Encountering an error when loading Godot project on CI after last update

### ISSUE THREAD

**Issue by Minimata** [MikeSchulze/gdUnit4#487]
### The used GdUnit4 version

4.3.0 (Pre Release/Master branch)

### The used Godot version

v4.2.2

### Operating System

Ubuntu container from the nektos/act runner handled by Gitea

### Describe the bug

Hi,
I opened Godot today after a 3-days hiatus to see that GdUnit suggested an update (with a new logo, yay!)
I installed the update and restarted Godot, no issues. I immediately pushed the update on my repo where the Gitea Action CI took over and this is where trouble begins.

The step that triggers the issue is as follows:
```yaml
      - name: Import resources and build solution
        run: |
          godot --headless --build-solutions --quit --import --path $PWD
```
So it simply launches godot to import all the necessary stuff before actually running the tests.

With the new version of GdUnit, this step now triggers the following error:
```
Loading GdUnit4 Plugin success
GdUnit4: Test server successfully started checked port: 31002
ERROR: Parent node is busy adding/removing children, `remove_child()` can't be called at this time. Consider using `remove_child.call_deferred(child)` instead.
   at: remove_child (scene/main/node.cpp:1449)
ERROR: Parent node is busy adding/removing children, `remove_child()` can't be called at this time. Consider using `remove_child.call_deferred(child)` instead.
   at: remove_child (scene/main/node.cpp:1449)
ERROR: Condition "data.parent" is true.
   at: ~Node (scene/main/node.cpp:[35](https://git.game-dev.space/minimata/Godot-CICDTests/actions/runs/72/jobs/1#jobstep-4-35)83)
Unload GdUnit4 Plugin success
Segmentation fault (core dumped)
```
My CI pipeline therefore stops right there and returns an error.

I have a few SCRIPT ERRORs beforehand regarding some missing png file and other UI stuff, I think it doesn't impact the actual issue so I won't add the whole logs to keep it as concise as possible but know that this is not the complete log. Also,`--verbose` doesn't add anything to these errors.

I tried to simply disable the plugin which actually solves this error but then another error appears when I try to export my game for Windows with the following command:
```yaml
      - name: Windows Build
        run: |
          mkdir -v -p build/windows
          godot --headless --verbose --build-solutions --import --export-release "Windows Desktop" build/windows/Windows.exe
```

This now triggers the following error (again not the whole logs but the useful information should live in here):
```
WARNING: Project export for preset "Windows Desktop" completed with warnings.
     at: _fs_changed (editor/editor_node.cpp:999)
reimport: end
ERROR: Attempt to disconnect a nonexistent connection from '<RefCounted#-9223370350828391021>'. Signal: 'gdunit_event', callable: 'Control(GdUnitConsole.gd)::_on_gdunit_event'.
   at: _disconnect (core/object/object.cpp:1420)
ERROR: Attempt to disconnect a nonexistent connection from '<RefCounted#-9223370350828391021>'. Signal: 'gdunit_message', callable: 'Control(GdUnitConsole.gd)::_on_gdunit_message'.
   at: _disconnect (core/object/object.cpp:1416)
ERROR: Attempt to disconnect a nonexistent connection from '<RefCounted#-9223370350828391021>'. Signal: 'gdunit_client_connected', callable: 'Control(GdUnitConsole.gd)::_on_gdunit_client_connected'.
   at: _disconnect (core/object/object.cpp:1420)
ERROR: Attempt to disconnect a nonexistent connection from '<RefCounted#-9223370350828391021>'. Signal: 'gdunit_client_disconnected', callable: 'Control(GdUnitConsole.gd)::_on_gdunit_client_disconnected'.
   at: _disconnect (core/object/object.cpp:1420)
EditorSettings: Save OK!
XR: Clearing primary interface
XR: Removed interface "Native mobile"
XR: Removed interface "OpenXR"
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.2.2.stable.official (15073afe3856abd2aa1622492fe50026c7d63dc1)
Dumping the backtrace. 
ERROR: FATAL: Index p_index = 1 is out of bounds (size() = 0).
   at: get (./core/templates/cowdata.h:158)
Illegal instruction (core dumped)
```

So I'm quite stuck because with plugin enabled I can't import my project correctly and with the plugin disabled I can't export my project at all.

My workaround as of now would be to manually install a previous version of GdUnit by going into the commit history of this repo (I haven't found a more straightforward way to install v4.2.5 or older but I might have missed it, I'm quite new to Godot). When I do this next week and when I find the most recent working version, I'll make sure to update this issue with the commit SHA or something. I hope this would help.

Thanks a lot for the plugin and for your help!

### Steps to Reproduce

1. Be on Godot 4.2.2 with Mono support
2. Install GdUnit4 through the AssetLib tab
3. Restart Godot
4. Accept and install the latest update
5. Commit and push the changed to a Gitea repository with actions enabled.
6. Actions take over by importing and building the project
7. Running the tests
8. Exporting a release

Issue happens at step 7 or 8 depending on plugin enabled or disabled respectively.

Below is the simplified yaml code for the gitea action:

```yaml  
jobs:
  ExportWindows:
    runs-on: ubuntu-latest
    container:
      image: barichello/godot-ci:4.2.2
    steps:
      - name: Cloning the repo
        uses: actions/checkout@v4
        with:
          lfs: true
      - name: Import resources and build solution
        run: |
          godot --headless --build-solutions --quit --import --path $PWD
      - name: Run tests
        run: |
          godot --headless --path "$PWD" -s -d addons/gdUnit4/bin/GdUnitCmdTool.gd -a ./test -rd ./test/reports --ignoreHeadlessMode
      - name: Windows Build
        run: |
          mkdir -v -p build/windows
          godot --headless --verbose --build-solutions --import --export-release "Windows Desktop" build/windows/Windows.exe
```

### Minimal reproduction project

Difficult to provide a project given that it works fine on my machine but fails on the CI container.

**Comment by MikeSchulze** [comment]
It's already fixed in master, I will provide a new release today
https://github.com/MikeSchulze/gdUnit4/issues/483

**Comment by MikeSchulze** [comment]
@Minimata version v4.3.1 is out, you can update

**Comment by Minimata** [comment]
You're the best, thank you so much. Will update ASAP.

**Comment by Minimata** [comment]
I just tried it but had the same issue. I'll completely uninstall and reinstall the plugin tomorrow with a clear head to see if it persists. If it does, I will fallback to manually downloading the repo's commit history until I find the most recent working version and share it with you.

**Comment by MikeSchulze** [comment]
@Minimata are you familiar with the gdunit4 action?
https://github.com/MikeSchulze/gdUnit4-action

**Comment by Minimata** [comment]
Ok so I completely uninstalled the plugin, reinstalled it from the AssetLib and enabled it (restarting Godot between every step), I'm on version 4.3.1 according to the plugin settings
![image](https://github.com/MikeSchulze/gdUnit4/assets/13940090/826e8edd-fd1a-4da8-b9b8-f40ab1d516c1)
And I still have the same error on project import.

I went dichotomy into the commit history and found out the last working commit is the following one:

0db9e5bd2cc2438e466749237f6081ebd44ee243 [Bump MikeSchulze/gdUnit4-action from 1.0.8 to 1.1.1](https://github.com/MikeSchulze/gdUnit4/commit/0db9e5bd2cc2438e466749237f6081ebd44ee243)

Which means the "remove_child" error is probably triggered by something somewhere in this commit:

0736ca3036dfab3ea0a3c0bb036e16c9cedd6cce [GD-322: Complete revision of the](https://github.com/MikeSchulze/gdUnit4/commit/0736ca3036dfab3ea0a3c0bb036e16c9cedd6cce) [GdUnit](https://github.com/MikeSchulze/gdUnit4/commit/0736ca3036dfab3ea0a3c0bb036e16c9cedd6cce) [inspector and addition of test discovery](https://github.com/MikeSchulze/gdUnit4/commit/0736ca3036dfab3ea0a3c0bb036e16c9cedd6cce)

It is interesting to note that the last working commit for the "remove_child" error is not the same as the last working commit for the "disconnect a non-existant connection" error.

The last working commit for the "disconnect" error (and my whole CI) is this one:

caa7eadd5ec643fece0e21f6c1938694d64717ed [Bump dorny/test-reporter from 1.9.0 to 1.9.1](https://github.com/MikeSchulze/gdUnit4/commit/caa7eadd5ec643fece0e21f6c1938694d64717ed)

which means the issue for this one lives in the following commit:

049424f59114ca82ff3f845a855aa94375b3c7d0 [GD-443: Fix errors when dock undock the](https://github.com/MikeSchulze/gdUnit4/commit/049424f59114ca82ff3f845a855aa94375b3c7d0) [gdunit](https://github.com/MikeSchulze/gdUnit4/commit/049424f59114ca82ff3f845a855aa94375b3c7d0) [inspector](https://github.com/MikeSchulze/gdUnit4/commit/049424f59114ca82ff3f845a855aa94375b3c7d0)

I know close to nothing on godot-specific dev and plugins so I don't really know where to look in these two commits pinpoint the actual issue but I will try.

**Comment by Minimata** [comment]
Alright, thanks for the info, I'll try to make your action work if the current issue is unsolveable.

Regarding the timeout, the `--import` godot argument is actually meant to do just that: prevent early exits but in a more straightforward way than the hardcoded timeout. Maybe it's buggy or doesn't take addons into account though, it was planned for 4.3 and cherry picked into 4.2.2 I believe, I'm going to try with a regular `--quit-after`.

Also, I'll try to setup a minimal reproduction example and come back to your right after.

Thanks for your help!

**Comment by Minimata** [comment]
Hi hello it's been a while.

I absolutely could not find the time to make a repro project then forgot about it and moved on to other stuff, I'm sorry. I'm back at it though!

First off, I tried your own GDUnit4 action on a bare-metal actions runner so it would not run into any Docker-in-docker issue and whatnot which is the issue I had last time I tried this action I believe. It ended up failing but without giving me a clear error message:

![image](https://github.com/user-attachments/assets/ea31fb63-37b0-4b33-94c0-10adfc8c82c8)

Here is the workflow that generates this problem:

```yaml
jobs:
  Tests:
    runs-on: godot
    steps:
      - name: Check out repository code
        uses: actions/checkout@v4
      - name: Import resources and build solution
        run: |
          godot --headless --editor --build-solutions --quit --import --path $PWD
      - name: run tests
        uses: MikeSchulze/gdUnit4-action@v1
        with:
          godot-version: '4.3.0'
          paths: 'res://tests'
          publish-report: false
          upload-report: false
```

I don't know if I misunderstood the usage of your action, I probably have, but I couldn't get it to work.

Then, I went back to trying to launch GDUnit4 tests using my previous method, i.e. using the baricello-ci container in Godot 4.3. It worked! Tests run fine and results can be uploaded as artifacts. Unfortunately, the addon generates an error afterwards when trying to build the game:

![image](https://github.com/user-attachments/assets/2cca672f-d464-4ebb-8f3e-3e20aa35324d)

Without the addon, this issue doesn't occur and the game builds for all platforms. Here is the workflow for this error:

```yaml
jobs:
  Tests:
    runs-on: godot
    container:
      image: barichello/godot-ci:4.3
    steps:
      - name: Check out repository code
        uses: actions/checkout@v4
      - name: Import resources and build solution
        run: |
          godot --headless --editor --build-solutions --quit --import --path $PWD
      - name: Run tests
        run: |
          godot --headless --path "$PWD" -s -d addons/gdUnit4/bin/GdUnitCmdTool.gd -a ./tests -rd ./tests/reports --ignoreHeadlessMode
      - name: Upload test results
        uses: actions/upload-artifact@v3
        if: always()
        with:
          name: Test results
          path: tests/reports
      - name: Windows Build
        run: |
          mkdir -v -p build/windows
          godot --headless --verbose --export-release "Windows Desktop" build/windows/windows.exe
```

The workaround I found was to have a first job for tests then a second job which forcefully remove the gdUnit4 folder from the project before building the game. A bit longer when ran sequentially but really not that much and it can be ran in parallel really easily.

So honestly this workaround works well for me but I figured that I would still mention this error to you so you're aware that it can happen. If you want I can create a new issue for either or both of these problems (the sudden stop of your action and the out of bounds index), or not because I'm fine with my workaround. Either way the original issue of the children removal is fixed as far as I can tell.

Thanks you for your help and work on this addon! Cheers!

**Comment by MikeSchulze** [comment]
You are using a very outdated version of the action the latest version is v1.1.1
https://github.com/MikeSchulze/gdUnit4-action/releases

update your WF to `uses: MikeSchulze/gdUnit4-action@v1.1.1`

**Comment by Minimata** [comment]
Ah, indeed. This and the godot version to `4.3` instead of `4.3.0` seem to work better. I still have an issue because it cannot find `xvfb-run` even though it's installed, but this is a problem with my VPS probably, I would have to investigate.

The build command still won't work with the addon installed though, but again, don't know if you want to investigate further or not.

**Comment by MikeSchulze** [comment]
> Ah, indeed. This and the godot version to `4.3` instead of `4.3.0` seem to work better. I still have an issue because it cannot find `xvfb-run` even though it's installed, but this is a problem with my VPS probably, I would have to investigate.
> 
> The build command still won't work with the addon installed though, but again, don't know if you want to investigate further or not.

to analyze the build issue, I need more detailed information
